compiler/rustc_span/src/lib.rs

```diff
@@ -100,19 +100,35 @@ impl SessionGlobals {
     }
 }
 
-pub fn with_session_globals<R>(edition: Edition, f: impl FnOnce() -> R) -> R {
+pub fn create_session_globals_then<R>(edition: Edition, f: impl FnOnce() -> R) -> R {
     let session_globals = SessionGlobals::new(edition);
+    assert!(!SESSION_GLOBALS.is_set(), "SESSION_GLOBALS should never be overwritten!");
     SESSION_GLOBALS.set(&session_globals, f)
 }
 
-pub fn with_default_session_globals<R>(f: impl FnOnce() -> R) -> R {
-    with_session_globals(edition::DEFAULT_EDITION, f)
+pub fn set_session_globals_then<R>(session_globals: &SessionGlobals, f: impl FnOnce() -> R) -> R {
+    #[cfg(not(parallel_compiler))]
```

Comment:
Can you add a FIXME explaining why this is necessary?

Reply:
Sure.